### ISSUE TITLE
Proposal: equivalent to `caf::json*` but for XML?

### ISSUE THREAD

**Issue by ethindp** [actor-framework/actor-framework#1922]
Though there are libraries that can do this, it would be neat if CAF supported it as it does for JSON. I know that JSON is more common for data interchange nowadays, but there are still times when one needs to work with XML data, and doing it dynamically is always annoying. I would do this but I just don't know how -- I'm not that experienced with C++ metaprogramming. Would this be something for CAF::core or perhaps another library?

**Comment by Neverlord** [comment]
We natively support JSON since it's the most common format. It's a good question where we should draw the line. There are other formats like YAML and INI/TOML that probably see more use nowadays than XML. XML is also a weird one, because of the attributes. For YAML, we could probably use a very similar data structure that we use for JSON. XML would be more work.

That's not to say we won't do XML at some point. We'll have to discuss this a bit, also more generally where we would place things like YAML and XML.

**Comment by ethindp** [comment]
I think that for XML a simple algorithm could be followed to distinguish between attributes and elements, even sticking within the bounds of type traits and concepts:

* If the field name of a struct is an [XSD 1.1 data type](https://www.w3.org/TR/xmlschema11-2), or an `std::optional` of any of the aforementioned types, or an `std::unique_ptr` or `std::shared_ptr` of the aforementioned types, then by definition it is an attribute
* Otherwise, it is an element.

The most significant problem is encoding all of the types, but I (think) a type mapping like the following might work:

| XSD data type | C++ equivalent |
| --- | --- |
| `anySimpleType` | `std::any` |
| `anyAtomicType` | `std::atomic<std::any>`? |
| `string` | `std::string` |
| `boolean` | `bool` |
| `decimal` | `long double`, `float128_t`, something like [cppdecimal](https://github.com/semihc/CppDecimal) |
| `float` | `float`, `float32_t` |
| `double` | `double`, `float64_t` |
| `duration` | `std::chrono::duration`? |
| `dateTime`, `time`, `date`, `gYearMonth`, `gYear`, `gMonthDay`, `gDay`, `gMonth` | `boost.datetime`, [date](https://github.com/HowardHinnant/date), or something else? |
| `hexBinary` | `std::vector<std::byte>` or `std::string`? |
| `base64Binary` | same as `hexBinary`? |
| `anyURI` | Something like the URL/URI types in [Ada](https://github.com/ada-url/ada)? |
| `QName` | `std::tuple<anyURI, NCName>`? |
| `NOTATION` | I'd say just error if we hit this type |
| `normalizedString` | `std::string` |
| `token` | `std::string` |
| `language` | `std::string` |
| `NMTOKEN` | `token` |
| `NMTOKENS` | `std::vector<NMTOKEN>` |
| `Name` | `std::string` |
| `NCName` | `std::string` |
| `ID` | `std::string` |
| `IDREF` | `std::string` |
| `IDREFS` | `std::vector<IDREF>` |
| `ENTITY` | `std::string` |
| `ENTITIES` | `std::vector<ENTITY>` |
| `integer` | `int` |
| `nonPositiveInteger` | `int` |
| `negativeInteger` | `int` |
| `long` | `long` |
| `int` | `int` |
| `short` | `short` |
| `byte` | `std::uint8_t`, `std::byte` |
| `nonNegativeInteger` | `unsigned int` |
| `unsignedLong` | `unsigned long` |
| `unsignedInt` | `unsigned int` |
| `unsignedShort` | `unsigned short` |
| `unsignedByte| ` |std::uint8_t`, `unsigned char`, `std::byte` |
| `positiveInteger` | `unsigned int| 
| `yearMonthDuration` | `std::chrono::duration` |
| `dayTimeDuration` | `std::chrono::duration` |
| `dateTimeStamp` | I... Am not sure about this one |

Obviously this isn't perfect and definitely needs some refinement, but I think something like this theoretically might work.

**Comment by ethindp** [comment]
Yeah so my mapping for `anySimpleType` and `anyAtomicType` is probably definitely wrong. I'm uncertain how to map those, perhaps `std::any` for both? Or maybe an `std::variant` for `anyAtomicType` or a concept?

**Comment by Neverlord** [comment]
That document is for schema definition. For an XML API, I believe we wouldn't need that unless we want to also implement validators, no?
